crypto/provider_core.c

```diff
@@ -1612,50 +1618,257 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
+            else
+                BIO_printf(b, "error getting value\n");
+            break;
+        case OSSL_PARAM_INTEGER:
+            if (OSSL_PARAM_get_int64(p, &i))
+                BIO_printf(b, "%lld\n", (long long int)i);
+            else
+                BIO_printf(b, "error getting value\n");
+            break;
+        case OSSL_PARAM_UTF8_PTR:
+            BIO_printf(b, "'%s'\n", *(char **)(p->data));
+            break;
+        case OSSL_PARAM_UTF8_STRING:
+            BIO_printf(b, "'%s'\n", (char *)p->data);
+            break;
+        case OSSL_PARAM_OCTET_PTR:
+        case OSSL_PARAM_OCTET_STRING:
+            BIO_printf(b, "<%zu bytes>\n", p->data_size);
+            break;
+# ifndef OPENSSL_SYS_UEFI
+        case OSSL_PARAM_REAL:
+            if (OSSL_PARAM_get_double(p, &d))
+                BIO_printf(b, "%f\n", d);
+            else
+                BIO_printf(b, "error getting value\n");
+            break;
+# endif
+        default:
+            BIO_printf(b, "unknown type (%u) of %zu bytes\n",
+                       p->data_type, p->data_size);
+            break;
+        }
+    }
+}
+#endif
+
+/**
+ * @brief Tears down the given provider.
+ *
+ * This function calls the `teardown` callback of the given provider to release
+ * any resources associated with it. The teardown is skipped if the callback is
+ * not defined or, in non-FIPS builds, if the provider is a child.
+ *
+ * @param prov Pointer to the OSSL_PROVIDER structure representing the provider.
+ *
+ * If tracing is enabled, a message is printed indicating that the teardown is
+ * being called.
+ */
 void ossl_provider_teardown(const OSSL_PROVIDER *prov)
 {
     if (prov->teardown != NULL
 #ifndef FIPS_MODULE
             && !prov->ischild
 #endif
-       )
+        ) {
+#ifndef FIPS_MODULE
+        OSSL_TRACE_BEGIN(PROVIDER) {
+            BIO_printf(trc_out, "(provider %s) calling teardown\n",
+                       ossl_provider_name(prov));
+        } OSSL_TRACE_END(PROVIDER);
+#endif
         prov->teardown(prov->provctx);
+    }
 }
 
+/**
+ * @brief Retrieves the parameters that can be obtained from a provider.
+ *
+ * This function calls the `gettable_params` callback of the given provider to
+ * get a list of parameters that can be retrieved.
+ *
+ * @param prov Pointer to the OSSL_PROVIDER structure representing the provider.
+ *
+ * @return Pointer to an array of OSSL_PARAM structures that represent the
+ *         gettable parameters, or NULL if the callback is not defined.
+ *
+ * If tracing is enabled, the gettable parameters are printed for debugging.
+ */
 const OSSL_PARAM *ossl_provider_gettable_params(const OSSL_PROVIDER *prov)
 {
-    return prov->gettable_params == NULL
-        ? NULL : prov->gettable_params(prov->provctx);
+    const OSSL_PARAM *ret = NULL;
+
+    if (prov->gettable_params != NULL) {
+        ret = prov->gettable_params(prov->provctx);
+#ifndef FIPS_MODULE
+        OSSL_TRACE_BEGIN(PROVIDER) {
```

Comment:
No trace is generated if provider doesn't implement "gettable_params". Intentional?

Reply:
It should trace that it does not exist IMHO ...